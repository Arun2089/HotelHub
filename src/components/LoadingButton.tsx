import { useState } from "react";
import { Button, CircularProgress, ButtonProps } from "@mui/material";

type LoadingButtonType = {
  buttonText: string;
  color?: ButtonProps["color"];
  variant?: "text" | "outlined" | "contained";
  className?: string;
  onSubmit: () => Promise<void> | void;
  size?: ButtonProps["size"];
  disabled?: boolean;
};

const LoadingButton = ({
  buttonText,
  color = "primary",
  variant = "contained",
  className = "",
  size = "large",
  disabled = false,
  onSubmit,
}: LoadingButtonType) => {
  const [loading, setLoading] = useState(false);

  const handleClick = async () => {
    setLoading(true);
    try {
      await new Promise((resolve) => setTimeout(resolve, 1000));
      await onSubmit();
    } finally {
      setLoading(false);
    }
  };

  return (
    <Button
      className={className}
      color={color}
      variant={variant}
      disabled={loading || disabled}
      onClick={handleClick}
      size={size}
      startIcon={loading && <CircularProgress size={16} />}
    >
      {buttonText}
    </Button>
  );
};

export default LoadingButton;
