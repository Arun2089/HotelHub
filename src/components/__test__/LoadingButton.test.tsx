import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import LoadingButton from "../LoadingButton";

const btnText = "Submit";
const mockOnSubmit = jest.fn();

describe("LoadingButton", () => {
  it("snapshot test", () => {
    const { asFragment } = render(
      <LoadingButton buttonText={btnText} onSubmit={mockOnSubmit} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render the button with the provided text", () => {
    render(<LoadingButton buttonText={btnText} onSubmit={mockOnSubmit} />);
    const buttonElement = screen.getByText(btnText);
    expect(buttonElement).toBeInTheDocument();
  });

  it("should call the onSubmit function when clicked", async () => {
    render(<LoadingButton buttonText={btnText} onSubmit={mockOnSubmit} />);
    const buttonElement = screen.getByText(btnText);
    fireEvent.click(buttonElement);
    await waitFor(() => {
      expect(mockOnSubmit).toHaveBeenCalled();
    });
  });

  it("should disable the button while loading", async () => {
    render(<LoadingButton buttonText={btnText} onSubmit={mockOnSubmit} />);
    const buttonElement = screen.getByText(btnText);
    fireEvent.click(buttonElement);
    expect(buttonElement).toBeDisabled();
    await waitFor(() => {
      expect(buttonElement).not.toBeDisabled();
    });
  });
});
