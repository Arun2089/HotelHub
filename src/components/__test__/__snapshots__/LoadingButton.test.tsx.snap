// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingButton snapshot test 1`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge css-jh47zj-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    Submit
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;
